### ISSUE TITLE
New version MaskComposite Error

### ISSUE THREAD

**Issue by h3clikejava** [comfyanonymous/ComfyUI#1608]
New ComfyUI Revision: 1496 [aeba1cc2] | Released on '2023-09-26'

![screenshot-20230927-125810](https://github.com/comfyanonymous/ComfyUI/assets/4688773/f5355cbf-078a-4c72-b6d7-a38ea88e34d8)

Error occurred when executing MaskComposite:

The size of tensor a (512) must match the size of tensor b (256) at non-singleton dimension 2

File "/home/h3c/Documents/ComfyUI/execution.py", line 152, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "/home/h3c/Documents/ComfyUI/execution.py", line 82, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "/home/h3c/Documents/ComfyUI/execution.py", line 75, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "/home/h3c/Documents/ComfyUI/comfy_extras/nodes_mask.py", line 247, in combine
output[:, top:bottom, left:right] = destination_portion * source_portion

The previous version was find; it became like this after upgrading to the new version.

**Comment by ltdrdata** [comment]
Did that work correctly in the previous version? It seems like it wouldn't have worked even in the previous version if the mask sizes were different.

**Comment by h3clikejava** [comment]
It's work fine in omfyUI Revision:  1485

**Comment by Kinglord** [comment]
I'm seeing this issue as well on the latest version. The error has to be with the node itself because when I feed to 2 of the exactly te same mask (identical in size and composition) it still throws this tensor error.

**Comment by rsamf** [comment]
I found where the problem lies in the code. I was having an issue of it myself, so I decided to do some digging.

It looks like the source and destination tensors become permutated, from (H, W) to (W, H), while the output tensor is still in (H, W). So they are incompatible if source or destination is non-square. I will try to make a PR to this repo.

**Comment by rsamf** [comment]
I just realized your situation is different @h3clikejava. My PR will only fix cases of non-square masks of equal shapes. In your case, you have 2 square masks of different shapes that you are trying to compose. What do you expect the behavior to be?

**Comment by Kinglord** [comment]
Thanks so much @rsamf - that will fix my issue for sure. In regards to @h3clikejava - I'm not sure what you're trying to do by combining 2 masks of different sizes either :D - you probably want to scale one, or do some kind of overlay prep before trying to use that node - I'm assuming you were trying to overlay a smaller mask on a larger one.

**Comment by h3clikejava** [comment]
I need make a mask like this. It's work fine in ComfyUI Revision: 1485. 
Now I have to convert it to an image first, just like the screenshot I provided.

![screenshot-20230929-133154](https://github.com/comfyanonymous/ComfyUI/assets/4688773/4f4a8682-beb6-4972-8adc-63c2b5bcd9f9)

**Comment by Duemellon** [comment]
I'm having this problem too for non-square masks.

**Comment by 0x473354** [comment]
https://github.com/comfyanonymous/ComfyUI/blob/0e763e880f5e838e7a1e3914444cae6790c48627/comfy_extras/nodes_mask.py#L243-L244 are slicing along the wrong dimensions. This change fixes the issue to the best of my knowledge:
```python
        source_portion = source[:, :visible_height, :visible_width]
        destination_portion = destination[:, top:bottom, left:right]
```

**Comment by rsamf** [comment]
@h3clikejava @Duemellon have you tried merging in the changes from the PR https://github.com/comfyanonymous/ComfyUI/pull/1619 ?

**Comment by BeastDD** [comment]
@rsamf I tried the changes of PR #1619 and 1 change fixed the issues I was having with MaskComposite. I just applied the change of line 243 to 244 , I did not apply the other changes as they seem to create an issue where sometime it would work and some other time it wouldn't  randomly creating the issue with tensors not having the right shape. So if you want this fix, only change line 243 & 244. in your installation. Thank you for your help this was incredibly  frustrating!

**Comment by Khazar123** [comment]
> https://github.com/comfyanonymous/ComfyUI/blob/0e763e880f5e838e7a1e3914444cae6790c48627/comfy_extras/nodes_mask.py#L243-L244
> 
> are slicing along the wrong dimensions. This change fixes the issue to the best of my knowledge:
> ```python
>         source_portion = source[:, :visible_height, :visible_width]
>         destination_portion = destination[:, top:bottom, left:right]
> ```

Thx for the tiny fix, sad that this has to sit for so long, with such a minor code difference.

**Comment by Khazar123** [comment]
@rsamf @0x473354 Getting this error when I change the line 244
`Error occurred when executing MaskComposite:

too many indices for tensor of dimension 2

File "C:\ComfyUI_windows_portable\ComfyUI\execution.py", line 152, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "C:\ComfyUI_windows_portable\ComfyUI\execution.py", line 82, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "C:\ComfyUI_windows_portable\ComfyUI\execution.py", line 75, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "C:\ComfyUI_windows_portable\ComfyUI\comfy_extras\nodes_mask.py", line 244, in combine
destination_portion = destination[:, top:bottom, left:right]`

When I keep 244 the same, but change 243, it works perfectly for me...
I'm using the Impact pack and WAS suite, I don't know if that changes anything.

**Comment by rsamf** [comment]
@BeastDD PR #1619 only changes lines 243 and 244 https://github.com/comfyanonymous/ComfyUI/pull/1619/files. Thanks for the comment.

@Khazar123 Can you find a stable case that always errors out when both lines 243 and 244 are changed.

**Comment by Khazar123** [comment]
> @BeastDD PR #1619 only changes lines 243 and 244 https://github.com/comfyanonymous/ComfyUI/pull/1619/files. Thanks for the comment.
> 
> @Khazar123 Can you find a stable case that always errors out when both lines 243 and 244 are changed.

Yep, this workflow will cause the error when both lines are changed:
[tensorError (3).txt](https://github.com/comfyanonymous/ComfyUI/files/12861943/tensorError.3.txt)

I disconnected anything unneeded to cause the error, all one needs is to input their checkpoint in the efficient loader and detection models in the mask section and hit the queue.

**Comment by 0x473354** [comment]
@Khazar123 @BeastDD The "SEGS to MASK (combined)" node outputs the mask as a 2D tensor (of shape [height, width]) but MaskComposite expects a 3D tensor (of shape [?, width height], might expect batch size as the zeroth dimension). MaskComposite already handles this edge case for source, but not for destination. You can fix this by changing
https://github.com/comfyanonymous/ComfyUI/blob/0e763e880f5e838e7a1e3914444cae6790c48627/comfy_extras/nodes_mask.py#L236 to:
```python
        destination = destination.reshape((-1, destination.shape[-2], destination.shape[-1]))
        output = destination.clone()
```
in combination with the previous fix.

**Comment by Khazar123** [comment]
@0x473354 that worked thank you! If source already handled that edge case fine, then I guess this change should flow into the PR? @rsamf
